// ecoledirecte.js/dist/accounts/Account.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Account = void 0;
const v3_1 = require("ecoledirecte-api-types/v3");
const functions_1 = require("../functions");
class Account {
    constructor(session) {
        this.__session = session;
        // Necessary checks
        if (!session.loginRes || v3_1.isFailure(session.loginRes))
            throw new Error("Account class must have valid connection");
        const mainAccount = functions_1.getMainAccount(session.loginRes.data.accounts);
        if (!session.token)
            throw new Error("Account class MUST have token");
        this.__account = mainAccount;
    }
    get token() {
        return this.__session.token;
    }
    set token(value) {
        this.__session.token = value;
    }
    get credentials() {
        return this.__session.credentials;
    }
    /**
     * @description EcoleDirecte unique Id
     */
    get edId() {
        return this.__account.id;
    }
}
exports.Account = Account;
