// ecoledirecte.js/dist/accounts/Teacher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Teacher = void 0;
const Account_1 = require("./Account");
const v3_1 = require("ecoledirecte-api-types/v3");
const functions_1 = require("../functions");
class Teacher extends Account_1.Account {
    constructor(session) {
        super(session);
        this.session = session;
        this.type = "teacher";
        const mainAccount = functions_1.getMainAccount(session.loginRes.data.accounts);
        if (!v3_1.isTeacherAccount(mainAccount))
            throw new Error("Teacher class's main account is wrong");
        if (!session.token)
            throw new Error("Account class MUST have token");
        this.account = mainAccount;
        this.token = session.token;
    }
    async getPhoto() {
        const r = await functions_1.fetchPhoto(this._raw);
        if (!r)
            return;
        const [buf, str] = r;
        this._photo = buf;
        this._photoUri = str;
        return buf;
    }
    get photo() {
        return {
            buffer: this._photo,
            uri: this._photoUri,
        };
    }
    get _raw() {
        return this.account;
    }
}
exports.Teacher = Teacher;
