// ecoledirecte-api-types/v3/routes.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Routes = exports.root = void 0;
const Util_1 = __importDefault(require("../util/Util"));
exports.root = "https://api.ecoledirecte.com";
exports.Routes = {
    //! LOGIN
    /**
     * Route for:
     * - POST /v3/login.awp
     */
    login() {
        return "/v3/login.awp";
    },
    //! STUDENT
    /**
     * Route for:
     * - POST /v3/login.awp
     */
    studentMailbox(studentId, params) {
        const defaultParams = {
            verbe: "getall",
            typeRecuperation: "received",
            orderBy: "date",
            order: "desc",
            page: 0,
            itemsPerPage: 0,
            onlyRead: "",
            query: "",
            idClasseur: 0,
        };
        const nParams = Util_1.default.mergeDefault(defaultParams, params);
        return Util_1.default.mergeParams(`/v3/eleves/${studentId}/messages.awp`, nParams);
    },
    /**
     * Route for:
     * - POST /v3/eleves/{studentId}/messages/{messageId}.awp
     */
    studentMessage(studentId, messageId, params) {
        const defaultParams = {
            verbe: "get",
            mode: "",
        };
        const nParams = Util_1.default.mergeDefault(defaultParams, params);
        return Util_1.default.mergeParams(`/v3/eleves/${studentId}/messages/${messageId}.awp`, nParams);
    },
    /**
     * Route for:
     * - POST /v3/eleves/{studentId}/cahierdetexte.awp
     */
    studentHomework(studentId, params) {
        const defaultParams = {
            verbe: "get",
        };
        const nParams = Util_1.default.mergeDefault(defaultParams, params);
        return Util_1.default.mergeParams(`/v3/Eleves/${studentId}/cahierdetexte.awp`, nParams);
    },
    /**
     * Route for:
     * - POST /v3/eleves/{studentId}/cahierdetexte/{date}.awp
     * @param date Must be of type YYYY-MM-DD
     */
    studentHomeworkDate(studentId, date, params) {
        const defaultParams = {
            verbe: "get",
        };
        const nParams = Util_1.default.mergeDefault(defaultParams, params);
        return Util_1.default.mergeParams(`/v3/Eleves/${studentId}/cahierdetexte/${date}.awp`, nParams);
    },
    /**
     * Route for:
     * - POST /v3/eleves/{studentId}/notes.awp
     */
    studentGrades(studentId) {
        const params = {
            verbe: "get",
        };
        return Util_1.default.mergeParams(`/v3/eleves/${studentId}/notes.awp`, params);
    },
    /**
     * Route for:
     * - POST /v3/eleves/{studentId}/timeline.awp
     */
    studentTimeline(studentId, params) {
        const defaultParams = {
            verbe: "get",
        };
        const nParams = Util_1.default.mergeDefault(defaultParams, params);
        return Util_1.default.mergeParams(`/v3/eleves/${studentId}/timeline.awp`, nParams);
    },
    /**
     * Route for:
     * - POST /v3/{accountType}/{accountId}/timelineAccueilCommun.awp
     */
    commonTimeline(accountType, accountId, params) {
        const defaultParams = {
            verbe: "get",
        };
        const nParams = Util_1.default.mergeDefault(defaultParams, params);
        return Util_1.default.mergeParams(`/v3/${accountType}/${accountId}/timelineAccueilCommun.awp`, nParams);
    },
};
