// ecoledirecte.js/dist/functions/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCommonTimeline = exports.getTimeline = exports.getGrades = exports.getMessages = exports.tickAssignement = exports.getTextbookPage = exports.fetchPhoto = exports.toISODate = exports.getMainAccount = exports.login = void 0;
var login_1 = require("./login");
Object.defineProperty(exports, "login", { enumerable: true, get: function () { return login_1.login; } });
Object.defineProperty(exports, "getMainAccount", { enumerable: true, get: function () { return login_1.getMainAccount; } });
var util_1 = require("./util");
Object.defineProperty(exports, "toISODate", { enumerable: true, get: function () { return util_1.toISODate; } });
var photo_1 = require("./photo");
Object.defineProperty(exports, "fetchPhoto", { enumerable: true, get: function () { return photo_1.fetchPhoto; } });
//! TEXTBOOK
var textbook_1 = require("./student/textbook");
Object.defineProperty(exports, "getTextbookPage", { enumerable: true, get: function () { return textbook_1.getTextbookPage; } });
var textbook_2 = require("./student/textbook");
Object.defineProperty(exports, "tickAssignement", { enumerable: true, get: function () { return textbook_2.tickAssignement; } });
//! MAILBOX
var mailbox_1 = require("./student/mailbox");
Object.defineProperty(exports, "getMessages", { enumerable: true, get: function () { return mailbox_1.getMessages; } });
//! NOTES
var grades_1 = require("./student/grades");
Object.defineProperty(exports, "getGrades", { enumerable: true, get: function () { return grades_1.getGrades; } });
//! TIMELINES
var timelines_1 = require("./student/timelines");
Object.defineProperty(exports, "getTimeline", { enumerable: true, get: function () { return timelines_1.getTimeline; } });
Object.defineProperty(exports, "getCommonTimeline", { enumerable: true, get: function () { return timelines_1.getCommonTimeline; } });
