// ecoledirecte.js/dist/functions/util.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EdHeaders = exports.makeRequest = exports.toISODate = void 0;
const node_fetch_1 = __importDefault(require("node-fetch"));
const v3_1 = require("ecoledirecte-api-types/v3");
const events_1 = __importDefault(require("../events"));
const errors_1 = require("../errors");
const events_2 = __importDefault(require("events"));
function toISODate(date) {
    const d = new Date(date);
    if (isNaN(d.getTime()))
        throw new Error("Invalid date");
    return [
        d.getFullYear(),
        (d.getMonth() + 1).toString().padStart(2, "0"),
        d.getDate().toString().padStart(2, "0"),
    ].join("-");
}
exports.toISODate = toISODate;
async function makeRequest(options = { method: "GET", url: "", guard: false }, context = {}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
) {
    const { method, url, body, guard } = options;
    const resListener = new events_2.default();
    function onRes(callback) {
        resListener.on("response", callback);
    }
    function offRes(callback) {
        resListener.off("response", callback);
    }
    events_1.default.emit("request", { method, url, body, context, onRes, offRes });
    const params = {
        method: method,
        headers: exports.EdHeaders,
    };
    if (method === "POST") {
        const urlencoded = new URLSearchParams();
        urlencoded.append("data", JSON.stringify(body));
        params.body = urlencoded;
    }
    const response = await node_fetch_1.default(url, params);
    const resBody = await response.json();
    resListener.emit("response", { response, body: resBody });
    if (guard && v3_1.isFailure(resBody))
        throw new errors_1.EcoleDirecteAPIError(resBody);
    return resBody;
}
exports.makeRequest = makeRequest;
exports.EdHeaders = {
    authority: "api.ecoledirecte.com",
    accept: "application/json, text/plain, */*",
    "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.88 Safari/537.36",
    "content-type": "application/x-www-form-urlencoded",
    origin: "https://www.ecoledirecte.com",
    "sec-fetch-site": "same-site",
    "sec-fetch-mode": "cors",
    "sec-fetch-dest": "empty",
    referer: "https://www.ecoledirecte.com/",
    "accept-language": "fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7",
};
